emscripten/native/ui/EmscriptenInput.cpp

```diff
@@ -0,0 +1,228 @@
+
+    long _pointer1 = -1;
+    long _pointer2 = -1;
+    bool isPressed = false;
+
+    long lastX = 0;
+    long lastY = 0;
+
+    long lastPressedX = 0;
+    long lastPressedY = 0;
+
+    long buttonTimestamp [3] = { -1, -1, -1 }; 
+    long lastDoubleClickTimestamp = -1;
+
+
+    EM_BOOL _emscripten_mouse_doubleClick(int eventType, const EmscriptenMouseEvent *mouseEvent, void *userData) {
+        MapView* mapView = (MapView*)userData;
+        if (mouseEvent->button == 0) {
+            mapView->onWheelEvent(4, mouseEvent->targetX * mapView->getScale(), mouseEvent->targetY * mapView->getScale());
+        } else if (mouseEvent->button == 2) {
+            mapView->onWheelEvent(-4, mouseEvent->targetX * mapView->getScale(), mouseEvent->targetY * mapView->getScale());
+        }
+        return true;
+    }
+
+    EM_BOOL _emscripten_mouse_start(int eventType, const EmscriptenMouseEvent *mouseEvent, void *userData) {
+        MapView* mapView = (MapView*)userData;
+        if (mouseEvent->buttons == 1) {
+            mapView->onInputEvent(NATIVE_ACTION_POINTER_1_DOWN, mouseEvent->targetX * mapView->getScale(), mouseEvent->targetY * mapView->getScale(), NATIVE_NO_COORDINATE, NATIVE_NO_COORDINATE);
+        }
+        isPressed = true;
+        return true;
+    }
+
+    EM_BOOL _emscripten_mouse_move(int eventType, const EmscriptenMouseEvent *mouseEvent, void *userData) {
+        MapView* mapView = (MapView*)userData;
+        lastX = mouseEvent->targetX * mapView->getScale();
+        lastY = mouseEvent->targetY * mapView->getScale();
+        if (isPressed) {
+            buttonTimestamp[0] = -1;
+            buttonTimestamp[1] = -1;
+            buttonTimestamp[2] = -1;
+            if (mouseEvent->buttons == 2 || (mouseEvent->buttons == 1 && (mouseEvent->ctrlKey || mouseEvent->shiftKey || mouseEvent->altKey || mouseEvent->metaKey))) {
+                if (lastPressedX != 0 && lastPressedY != 0) {
+                    long diffX = lastX - lastPressedX;
+                    long diffY = lastY - lastPressedY;
+
+                    int x2 = mapView->getCanvasWidth() / 2;
+                    int y2 = mapView->getCanvasHeight() / 2;
+
+                    if (lastY > y2) diffX *= -1;
+
+                    mapView->rotate(diffX / 1.5, 0);
+                    mapView->tilt(diffY, 0);
+                } 
+                
+                lastPressedX = lastX;
+                lastPressedY = lastY;
+            } else if (mouseEvent->buttons == 1) {
+                mapView->onInputEvent(NATIVE_ACTION_MOVE, mouseEvent->targetX * mapView->getScale(), mouseEvent->targetY * mapView->getScale(), NATIVE_NO_COORDINATE, NATIVE_NO_COORDINATE);
+            }
+        }
+        
+        return true;
+    }
+
+    EM_BOOL _emscripten_mouse_end(int eventType, const EmscriptenMouseEvent *mouseEvent, void *userData) {
+        MapView* mapView = (MapView*)userData;
+        if (mouseEvent->timestamp - buttonTimestamp[mouseEvent->button] < 300) {
```

Comment:
Is keeping track of event timestamps needed here? `TouchHandler` instance should do this automatically.

Reply:
I deleted it.

I want to zoom in with double mouse left click or double single finger tap and zoom out with double mouse right click or double multi finger tap. TouchHandler instance does this automatically for only zoom in. Maybe we will add it for zoom out in the future.